polynote-kernel/src/main/scala/polynote/kernel/ScalaCompiler.scala

```diff
@@ -178,7 +187,14 @@ class ScalaCompiler private (
     List(treeBuilder.scalaDot(typeNames.AnyRef)), noSelfType, stats
   )
 
-  private def template(stats: Tree*): Template = template(stats.toList)
+  // the default implementation of atPos does some additional junk that messes up things (and isn't threadsafe or nesting-safe)
```

Comment:
To expound on this, the compiler's `atPos` has some issues:
- The `Global` instance has one instance of `PosAssigner`, which has a mutable variable `pos`. This mutable variable is used to store the default position we're going to assign.
   - This means that if you call `atPos` from within another call to `atPos` (say, you're point-free building some positioned trees) then `pos` will get clobbered.
   - This also means if another thread calls `atPos` for the same `Global` then it will clobber `pos`.
- If a tree is positioned, it doesn't check if the subtrees are positioned (so you may end up with unpositioned subtrees)
- For unpositioned subtrees which have positioned children, it does *not* use the position you specified, but rather generates a position based on the positioned children in order to "wrap" them gently in a nice loose-fitting warm position. This interferes dramatically with other presentation compiler functionality, because then (for example) when the code is a one-liner it puts the class's tree at the same position as the one liner even if there's linebreaks around the one liner.